### ISSUE TITLE
warning: skipping duplicate package `case` found at `~\.cargo\git\checkouts\cargo-..`

### ISSUE THREAD

**Issue by aminya** [rust-lang/cargo#10752]
### Problem

When using `cargo install --git`  to install a package like cargo itself, a lot of warnings are emitted about duplicated packages. It seems that these packages are only used for the tests, and not sure if the warnings affect the final build.

```
warning: skipping duplicate package `cargo-list-test-fixture` found at `~\.cargo\git\checkouts\cargo-..`
...
```

It seems that this is caused by this commit:
https://github.com/rust-lang/cargo/commit/e903f7dfc0e50fe3e11dcbc5456094c902fce6ce

![image](https://user-images.githubusercontent.com/16418197/173324225-e53f6ade-8498-4e20-964f-b2bf7bc42343.png)

### Steps

Try to build cargo with itself:
```
cargo install --git https://github.com/rust-lang/cargo
```

### Version

```text
cargo 1.63.0-nightly (38472bc19 2022-05-31)
```

**Comment by epage** [comment]
Interesting.  cargo is searching recursively through cargo's git repo and finding all of the test cases we have in the repo and reporting a warning for those as we duplicate the manifest names.   These should not even be given as options for install for the warning to be activated.

Besides telling the user to run `cargo install --git https://github.com/rust-lang/cargo cargo` I wonder if there is some way we can allow repos to exclude directories from being searched for manifests.

**Comment by epage** [comment]
We discussed this in the cargo team meeting and came up with a short term and long-term idea

Short term is to see if we can make the warning only appear when the package is `publish = true`

There is still the wider problem if packages being looked up that are internal.  Ideas
- a breadcrumb file that says "stop recursively walking for path/git sources"
- a new manifest key that says "stop recursing and use workspace lookup rules"

**Comment by epage** [comment]
@danilhendrasr since you worked on the warnings in #10701, I thought I'd first check if you would be interested in seeing if the warning could be restricted to packages with `publish = false` to reduce noise from the warning.

**Comment by danilhendrasr** [comment]
Sure, I'll take a stab at it.

**Comment by epage** [comment]
Hmm, thinking about this, i suspect #10767 only gave people the tool to reduce the warnings but cargo still needs to opt-in to that tool, so going ahead and re-opening for now.

**Comment by ehuss** [comment]
Would it be possible to restrict the warning to situations where the package is actually used?  That is, when doing `cargo install --git` or having a git dependency, only issue a warning if that package was actually part of the selected packages?  I suspect that won't be easy since `read_packages` is pretty low-level, but I think from a user experience, that is what would be expected.  Warning about unrelated packages seems to be a false-positive.  Also, forcing people to use a workaround like `publish=false` does not seem ideal.

I would also expect this to actually be an error in the long-run.  In the original issue (#10669), this leads to non-deterministic behavior.  A warning just says "something's broken" and then proceeds doing something random, which doesn't seem good.

Also, the current warning does not seem to be very helpful (particularly for an end-user who may have no idea where this is coming from, or is not well-versed in cargo). There is no actionable advice given to the user, or any context as to how cargo got into this bind. I would expect in the situation with a git dependency that has a duplicate, it would say something like:

```
error: multiple packages named `foo` found in git repository 
https://github.com/rust-lang/example for dependency `foo` specified in `path/to/Cargo.toml`
    First instance found at `path/in/repo/Cargo.toml`
    Second instance found at `path/in/repo/other-foo/Cargo.toml`
    A git repository must only contain a single package with the name `foo`.
```

With similar errors for `paths` overrides and `install --git`.

(Without the ability to restrict a git repo to a specific path, the solution of being unique isn't great, but should be fine in most situations.)

**Comment by epage** [comment]
> Would it be possible to restrict the warning to situations where the package is actually used? That is, when doing cargo install --git or having a git dependency, only issue a warning if that package was actually part of the selected packages? I suspect that won't be easy since read_packages is pretty low-level, but I think from a user experience, that is what would be expected. Warning about unrelated packages seems to be a false-positive. Also, forcing people to use a workaround like publish=false does not seem ideal.

The big issue with the case in question is that the user is doing 
```console
$ cargo install --git https://github.com/rust-lang/cargo
```
and not
```console
$ cargo install --git https://github.com/rust-lang/cargo -p cargo
```

The most we can filter it down is by whether the package has an installable artifact which I think would be an additional useful thing to do.

Overall, this still leaves the big hole of `cargo install` installing everything in the source regardless of the source author's intent.

**Comment by weihanglo** [comment]
Should we reconsider redirecting error to a log file? Or at minimal, Cargo can log some parts of the error that is not meaingful to end users. We then can provide a way to retrieve a verbose error report.

The workflow might look like this,

```
warning: error: multiple packages with the same name found in git repository 
This may cause non-deterministic result of the compilation.
Run `cargo report errors --id 168` to see the error report.
```

**Comment by spookyvision** [comment]
would it be possible to stop traversing at a depth where the requested package is found?
In my scenario:
- repository for library crate is https://github.com/spookyvision/embedded-graphics-web-simulator.git
- the actual library crate name is in the root and thus 
- it is not a workspace
- cargo however traverses into `examples/` (not a package) and finds two separate subdirectories there (`basic`, `animated`). Those are workspaces which each contain a package named `run-wasm`.
- adding this git repo as a dependency in another project with `embedded-graphics-web-simulator = { git = "https://github.com/spookyvision/embedded-graphics-web-simulator.git" }` leads to "warning: skipping duplicate package `run-wasm`". But as far as the user is concerned, the traversal into the repo depths was never necessary, they just requested the library in the root, which can be unambiguously resolved

**Comment by weihanglo** [comment]
FWIW, there is an issue proposing Cargo to have some controls over its own warnings: https://github.com/rust-lang/cargo/issues/12235

**Comment by PaulDance** [comment]
Just mentioning another use case: externally consuming `cargo-test-support` and `cargo-test-macro` in order to test a Cargo sub-command project.

https://github.com/rust-lang/cargo/issues/10752#issuecomment-1162678408 seems to me to be the ideal way to fix this particular warning. It would indeed restrict the analysis to what is expected by the user: what they end up using. Furthermore, the warning is issued for all build profiles, for example when the dependency is a `dev-` one and only `cargo build` is used, making it even more counter-intuitive and fixable by what is proposed above.

**Comment by weihanglo** [comment]
> Just mentioning another use case: externally consuming `cargo-test-support` and `cargo-test-macro` in order to test a Cargo sub-command project.

A solution specifically for these two packages: publish them 😆.

**Comment by PaulDance** [comment]
Yes, I thought about that too! It does however seem a bit beyond my acceptability knowledge and the current issue's specific topic: to discuss separately?

**Comment by weihanglo** [comment]
@PaulDance <https://github.com/rust-lang/cargo/issues/10147>

**Comment by PaulDance** [comment]
Thanks! -- sorry for not even having thought of looking it up.
